Symbolisches 84x48 Display

You69Man opened the pull request:
Das Ziel dieses PRs ist eine weitere Verbesserung der Display_Mono Anzeigen. 
Der PR besteht aus 2 commits:

Der Zweck des ersten commits (auch als eigener branch verfügbar: https://github.com/You69Man/ahoy/tree/feature/refactor_display_class_PR)  ist es, die Display Plugins besser von den restlichen Ahoy Datenstrukturen zu entkoppeln. Hierzu wird eine neue Übergabestruktur **DisplayData** eingeführt, welche von der Basis Display Klasse befüllt wird, und den einzelnen Plugins übergeben wird. Die Basis Display Klasse erhält für diesen Zweck auch Zugriff auf das IApp Interface, um von dort auf weitere möglicherweise anzueigende Daten zuzugreifen. Die einzelnen Display Plugins benötigen auf diese Art weniger Wissen über aktuelle Systemstrukturen und -Zustände, was die Schnittstellen und die Wartbarkeit vereinfacht. In der Display_Mono Klasse werden einige allen Plugins gemeinsame Methoden zusammengefasst, ebenfalls für bessere Wartbarkeit und um Code zu sparen. Nebenbei werden von dem Commit ein paar kleinere Einrückungs- und Schreibfehler korrigiert.

Der zweite Commit enthält neue Anzeige features für das 84x48 plugin. Um die geringe Pixeldichte so gut wie möglich zu nutzen wird Text an einigen Stellen durch Symbole ersetzt. Hierzu werden U8G2 Fonts direkt in das Ahoy Repo eingebunden, wo sie zum einen editiert werden können, und zum anderen durch Symbole ergänzt werden können. Dadurch werden gleich auch einige Schönheitsfehler des bestehenden 5x8 Font behoben, wie z.B. der Punkt, welcher zu einem + verunstaltet war, oder der 0er, welcher vorher viel zu schmal war. Die direkte Einbindung der Fonts hat praktisch keine negativen Konsequenzen auf die Flash Auslastung.

Zu den neuen features zählen:
- WiFi und MQTT Status über Symbol in rechter, unterer Ecke
- WiFi RSSI über Balken am rechten Rand
- NRF Modul Status (connected) über Symbol in linker unterer Ecke
- NRF RSSI über Balken am linken Rand (dzt noch eher statisch, sobald NRF RSSI (Ersatz-)Werte verfügbar, jedoch einfach über die DisplayData struktur anzubinden)
- Anzeige der Anzahl von offline und online Invertern (Sonne, Mond Symbole)
https://raw.githubusercontent.com/You69Man/ahoy/feature/display_84x48_symbolic_PR/pics/PXL_20230824_204200660.jpg

Falls gefällt, bitte mergen! 

PS: Es gibt auch bereits eine Implementierung für OLED 128x64:
https://raw.githubusercontent.com/You69Man/ahoy/feature/display_84x48_symbolic_PR/pics/PXL_20230901_061927908.jpg
Das doch recht volle und randfüllende Layout hat allerdings die Konsequenz, dass der "Bildschirmschoner" dort nicht mehr verwendet werden kann, was dazu führen würde, dass die OLEDs noch schnelle einbrennen. Daher lasse ich dort mal die Finger davon, ich möchte nicht Schuld für defekte Displays sein. :)

knickohr commented:
Kaputt werden die OLEDs nicht gehen, aber mit der Zeit werden halt die immer beleuchteten Pixel mehr und mehr verblassen. Vor allem das weiße Display, gefolgt vom gelben ist da sehr anfällig. Das blaue scheint dagegen sehr resistent zu sein. Andere Farben habe ich noch nicht getestet. Im Prinzip reicht es ja, das im Kreis herum um ein einziges Pixel zu verschieben, nicht wie es momentan ist mit bis zu 7 ? Eine Alternative wäre das Ganze als „Laufschrift“ zu gestalten. Sieht allerdings dann gewöhnungsbedürftig aus. Das wäre allerdings die optimale Lösung, da so jeder Pixel gleichmäßig bestrahlt wird und somit das Display überall gleich altert.

DanielR92 commented on src/plugins/Display/Display_Mono_128X64.h:
```diff
@@ -9,93 +9,81 @@
-            mDisplay->clearBuffer();
-            mDisplay->setContrast(mLuminance);
-            printText("AHOY!", 0, 35);
+            printText("AHoy!", 0, 35);
```
Wurde diese Zeile extra "umgeschrieben"? **AH**oy

You69Man replied:
Ja ich dachte eigentlich, in dieser Form stünde es auch auf der Homepage, aber ich sehe gerade, das war ein Irrtum. Ahoy! wäre korrekt.

DanielR92 commented on src/plugins/Display/Display_data.h:
```diff
@@ -0,0 +1,22 @@
+        bool RadioSymbol = false;       // indicate if radio module is connecting and working
+        bool MQTTSymbol = false;        // indicate if MQTT is connected
+        int8_t WifiRSSI=SCHAR_MIN;      // indicate RSSI value for WiFi
+        int8_t RadioRSSI=SCHAR_MIN;     // indicate RSSI value for radio
```
@lumapu Ist es mit Ahoy möglich NRF und CMT gleichzeitig zu betreiben? 
Dann müsste man hier nähmlich noch was hinzufügen?

DanielR92 commented:
Ich würde hier das ganze noch sauber trennen. Habe gesehen das auch andere "tools" im PR drinnen sind. "tools/fonts/Fony/Fony.exe" zum Beispiel. Das ist zum Font erstellen, jedoch gehört nicht in einem PR mit dazu.

Sonst sehr schön.

You69Man commented:
fony.exe kann man natürlich auch wieder rausnehmen, falls es stört, aber nachdem es schon einen Tools folder gibt, hätte man hier gleich die richtige Version zum editieren der Fonts dabei. Die Sourcen der Fonts mit den Symbolen ( extra für Ahoy! designed:) ) würde ich auf jeden Fall drin lassen. Es ist auch eine kleine batch Datei dabei, um die Fonts nach dem Editieren in den entsprechenden C Code umzuwandeln. Sonst müsste man sich das alles wieder zusammensuchen, falls man noch einmal was mit Fonts oder Symbolen machen möchte.

You69Man commented:
> Kaputt werden die OLEDs nicht gehen

Ich habe schon einmal so ein OLED weggeworfen nachdem es nach ca. zwei Jahren Dauerbetrieb praktisch unleserlich geworden ist. Ich glaube es war ein blaues, kann mich hier aber auch irren, da ich verschiedenste in meiner Bastelkiste habe. Ich habe bei meinen OLED Projekten praktisch immer den einen oder anderen Taster dabei, für eine Menüführung, damit wird dann auch das Display aktiviert, und nach 1 Minute Inaktivität wird das OLED dann wieder ausgeschaltet. So gibt es kein Problem. Das Pixel-herumschieben als Bildschirmschoner überzeugt mich weniger. Erstens verliert man die Randpixel (und somit Auflösung und Displaygröße), und andererseit ist trotzdem nie ausgeschlossen, dass nicht einem eingeschalteten Pixel wieder ein eingeschalteten Pixel nachgeschoben wird (z.b. bei fetten Fonts, Linien Symbolen). Dann hilft das alles nichts.

Jedenfalls gibt's diese Probleme zum Glück beim 84x48 LCD nicht. Da brennt nichts ein, und daher braucht's da auch keinen Bildschirmschoner. Daher bevorzuge ich bei diesem Projekt das Nokia 😀

knickohr commented:
An meinem Drucker habe ich ein weißes zum Wiegen der Filamentspulen, daraus kann man die verbleibende Länge berechnen, was übrigens hervorragend funktioniert. Trotz einem Font der nicht immer die gleichen Pixel aktiviert ist es nach gut einem Jahr schon fast blind 😕

Im Auto habe ich ein gelb/blaues, gleicher Font, gleicher Aufbau, zur Messung der Innen- und Außentemperatur. Das ist da jetzt schon 10 Jahre drin und zeigt keinerlei Ermüdungserscheinungen.

korrekt, das Pixelzittern ist eine Notlösung. Natürlich ist ein Abdimmen nach beispielsweise 1 Minute Optimsl. Aber es setzt voraus das irgendwo ein Taster ist, wo man es wieder aktivieren kann. beim ES32 sicher kein Problem, aber beim 8266er geht’s schon los wegen den fehlenden Pins,

Das war übrigens schon mehrmals gefordert, hat nur nie jemand umgesetzt. Damit könnte man dann auch beim Nokia die Hintergrundbeleuchtung aktivieren. 2 verfluchte Pins, die der 8266 nicht mehr hat 😞

You69Man commented:
>
> Das war übrigens schon mehrmals gefordert, hat nur nie jemand umgesetzt....2 verfluchte Pins, die der 8266 nicht mehr hat 😞

Wieso zwei Pins? Ein Taster wäre doch schon Mal besser als keiner. Was ist denn mit dem Ain0? Ist der nicht noch frei? Da könnte man sogar gleich mehrere Tasten dran machen...

knickohr commented:
Einen für den Taster und einen für die Hintergrundbeleuchtung des Nokia.

lumapu commented:
ich werde mir den PR anschauen und bin auch gewillt ihn zu mergen. Für heute ist es aber zu viel Content 😉

knickohr commented:
Du schwächelst 😂

lumapu commented:
@knickohr: stimmt, waren nur 11 Änderungen, die ich gerade gepusht habe 😅

knickohr commented:
@lumapu 
Ich lese gerade 😉

Offtopic :
Was bedeutet das ?  „ send loop skip disabled inverters fix“

lumapu commented:
> Offtopic : Was bedeutet das ? „ send loop skip disabled inverters fix“

Wenn man mehrere Inverter eingetragen hat, diese aber 'disabled' hat, dann wurde die Schleife trotzdem für diesen Inverter gezählt. Kleines Beispiel:
* 2 Inverter eingetragen, einer disabled
* Interval 10s

Früher wurde der einzig verbleibende Inverter nur alle 20s abgefragt, jetzt alle 10s.

PS: Jetzt wo ich meine Commit-Message so lese, merke ich dass die Formulierung mehr als holprig ist 😂

knickohr commented:
Ich hab’s befürchtet 😵

Wirft mein Timing bei mehreren DTUs durcheinander. Aber egal, man kann es ja entsprechend der Intervallzeit wieder gerade biegen.

You69Man commented:
> ich werde mir den PR anschauen und bin auch gewillt ihn zu mergen. Für heute ist es aber zu viel Content 😉

Super, danke 😊 Falls ich noch was ändern soll bitte im Info.

You69Man committed:
refactor display class

Display class gets access to App interface and HmRadio
Display class prepares relevant online information for displays in new DisplayData structure
Displays get pointer to DisplayData and only use information received from there

You69Man committed:
add fonts with symbols to ahoy project and add WLAN RSSI bar to 84x48

You69Man commented:
Habe den PR hochgezogen auf 0.7.48, da sich mittlerweile ein paar kleine merge conflicts ergeben haben. 
Bei der Gelegenheit auch das "AHoy!" auf "Ahoy!" geändert.
Falls ich noch was ändern soll, bitte einfach sagen. Falls ihr was weglassen wollt (die Font-Tools z.B.), bitte gerne rauslöschen.

lumapu commented:
wird heute gemerged. Habe die `exe` entfertn, evtl. muss hier noch ein Hinweis rein, wo man diese im Notfall findet. Toll wäre das natürlich als Python-Script, da versionierbarer Source-Code.
Kann man die Fonts in `Display_Mono.h` auch als Hex einfügen - finde das sieht dann sauberer aus, als dieser Datenwust 😉

You69Man commented:
Danke fürs Mergen! Die Fonts sind im gleichen Format, wie es auch die originalen U8G2 Fonts sind. Falls sie im Display_Mono zu sehr als Fremdkörper wirken, könnte man sie natürlich in ein eigenes .h File verschieben.

lumapu commented:
sehr gerne.

Ok, dann lassen wir es so, wenn es mich in Zukunft stört werde ich es auslagern.
